### PR TITLE
docs(mcp): spec install→auth handoff interaction logic

### PR DESCRIPTION
**Pull request by xfgong**
## Summary

After the four-layer MCP refactor merged, clicking **Install** in admin or workspace UI creates a `pending` install row and stops — no OAuth window, no static-credential form, no nudge. The catalog/server/override stack was removed (commit `243e6396`) and its OAuth start/callback handlers went with it; the four-layer replacements landed as `501` stubs (see `admin_mcp.py:373-389`, `ws_mcp.py:396-414`, `ws_mcp.py:499-517`) and the front end never learned to ask the user to authorize.

This PR adds **interaction-logic spec only**. Back-end OAuth wiring and front-end implementation land in follow-up PRs.

## What's in the spec

- **§3 Authentication action band** — a five-state band rendered inside the existing detail panel (`ready / needs-action / awaiting-others / oauth-in-flight / error`). One new UI surface, no redesigns.
- **§4 Caller authority matrix** — pure function of `required_grant_scope × caller role`. Decides which scope's action the UI surfaces.
- **§5 End-to-end flows** — text diagrams for member-installs-user-policy, admin-installs-workspace-policy, org-admin-installs-org-policy, and static-token install.
- **§5.5 OAuth pop-up controller** — `runOAuthFlow` contract: `window.open` + `BroadcastChannel('cubebox-mcp-oauth')` + 90 s timeout + `child.closed` cancellation polling.
- **§5.6 `/oauth/mcp/return` page** — typed `postMessage` over the channel, then `window.close()`.
- **§6 Back-end contract** — what the UI requires (`oauth/start` shape, callback redirect params, effective-state `reason` field).
- **§7 Edge cases** — popup blocked, state expired, AS denial, parallel admin authorizes, parent tab closed, policy flipped mid-flow.

## Design calls explicitly made by this spec

- **Two-step UX, not transactional install+auth.** Install and grant remain independent records so re-auth doesn't disturb the install row.
- **Action band lives inside the detail panel**, not in a modal. Matches `setMode('detail')` flow that already happens after install.
- **90 s OAuth pop-up timeout** (not 30 s, not 5 min) — accommodates MFA, fails fast on closed tabs.
- **`BroadcastChannel`** over `window.opener.postMessage` for resilience against sandboxed redirects.

## Out of scope

- Back-end OAuth start/callback implementation.
- Front-end implementation.
- Live updates (band auto-refresh when another tab authorizes).
- Notification system (`Notify` button is a layout placeholder).

## Review request

@codex — please review the interaction logic for correctness, completeness, and any contract gaps with the four-layer data model. Specifically scrutinize:

- §4 matrix vs. effective-state decision table from `docs/superpowers/plans/2026-05-16-mcp-management-four-layer.md` Task 5.
- §5.5/§5.6 pop-up + return-page contract — does the state/install_id round-trip survive same-site cookie restrictions?
- §7 edge cases — anything missing that the four-layer model can produce?

## Test plan

- [ ] Spec compiles in `docs/superpowers/specs/` directory layout
- [ ] @codex review pass with no actionable comments

### CODE REVIEW

File: docs/superpowers/specs/2026-05-16-mcp-install-auth-handoff-spec.md [hunk 1]
```diff
@@ -0,0 +1,462 @@
...
+ │  [ ••••••••••••••••••••••••••••••••••• ]  [ Save credential ]        │
+ └──────────────────────────────────────────────────────────────────────┘
+```
+
+- Shown when `connector.credential_availability === 'missing'` AND the caller
+  has authority to create the grant at `required_grant_scope` (see §4).
+- `<reason copy>` translates `connector.reason`:
+  - `pending_oauth` → "Authorization was started but not finished."
+  - `missing_grant` → "No credential on file yet."
+  - `expired_oauth` → "Your previous authorization expired."
```
[the remark this answers is not in the document]
Comment (xfgong) [reply docs/superpowers/specs/2026-05-16-mcp-install-auth-handoff-spec.md]: Fixed in a06afcbe6ed7d67f70c3ff224a181e6206094fe1 — see commit message for the three corrections.

File: docs/superpowers/specs/2026-05-16-mcp-install-auth-handoff-spec.md [hunk 2]
```diff
@@ -0,0 +1,462 @@
...
+
+```
+ ┌──────────────────────────────────────────────────────────────────────┐
+ │  ✓  Ready  ·  credential from <source>                               │
+ │                                                       [Disconnect ▾] │
+ └──────────────────────────────────────────────────────────────────────┘
+```
+
+- Shown when `connector.usable === true` AND `connector.credential_source !=
+  null`.
```
[the remark this answers is not in the document]
Comment (xfgong) [reply docs/superpowers/specs/2026-05-16-mcp-install-auth-handoff-spec.md]: Fixed in a06afcbe6ed7d67f70c3ff224a181e6206094fe1 — see commit message for the three corrections.

File: docs/superpowers/specs/2026-05-16-mcp-install-auth-handoff-spec.md [hunk 3]
```diff
@@ -0,0 +1,462 @@
...
+  that lens — it surfaces and creates the workspace grant for whichever
+  workspace the admin is currently viewing. The API call still goes to the
+  workspace-scoped endpoint (`/api/v1/ws/{lensWsId}/mcp/installs/{id}/grants/workspace[…]`),
+  not to an admin alias — there is no admin alias for workspace grants and
+  this spec does not add one.
+- An admin viewing an org-policy install still sees the `org` action band on
+  the org-wide install row regardless of lens.
+- If the org-policy install also has *user* policy override for the admin's
+  own membership (rare but legal), only the highest-priority missing scope's
+  band is shown. Priority: `org > workspace > user`.
```
[the remark this answers is not in the document]
Comment (xfgong) [reply docs/superpowers/specs/2026-05-16-mcp-install-auth-handoff-spec.md]: Fixed in a06afcbe6ed7d67f70c3ff224a181e6206094fe1 — see commit message for the three corrections.

**Comment by xfgong** [comment]
@codex review — applied three fixes (see a06afcbe). Please verify the spec is now consistent with backend/cubebox/mcp/effective.py reason tokens and the single-effective-policy resolution.

File: docs/superpowers/specs/2026-05-16-mcp-install-auth-handoff-spec.md [hunk 4]
```diff
@@ -0,0 +1,491 @@
...
+ │  ── or ──                                                             │
+ │  Static token                                                         │
+ │  [ ••••••••••••••••••••••••••••••••••• ]  [ Save credential ]        │
+ └──────────────────────────────────────────────────────────────────────┘
+```
+
+- Shown when `connector.credential_availability === 'missing'` AND the caller
+  has authority to create the grant at `required_grant_scope` (see §4).
+- `<reason copy>` translates `connector.reason` using the actual reason
+  tokens emitted by `backend/cubebox/mcp/effective.py` (`MCPEffectiveReason`
```
[the remark this answers is not in the document]
Comment (xfgong) [reply docs/superpowers/specs/2026-05-16-mcp-install-auth-handoff-spec.md]: Fixed in a747c2da90886f1d1ae31dcbe3619e0387b09621 — see commit message.

File: docs/superpowers/specs/2026-05-16-mcp-install-auth-handoff-spec.md [hunk 5]
```diff
@@ -0,0 +1,491 @@
...
+  otherwise `install.default_credential_policy` (see
+  `backend/cubebox/mcp/effective.py` rules 3-4). There is no simultaneous
+  multi-scope-missing case to disambiguate — exactly one scope is required
+  per (install, workspace) pair. If a workspace overrides an org-policy
+  install down to `user`, the band displayed on the admin's lens row for
+  that workspace becomes the `user` (per-admin) variant, not `org`.
+
+## 5. End-to-end interaction flows
+
+### 5.1 Workspace member installs a `user`-policy connector (e.g. Linear)
```
[the remark this answers is not in the document]
Comment (xfgong) [reply docs/superpowers/specs/2026-05-16-mcp-install-auth-handoff-spec.md]: Fixed in a747c2da90886f1d1ae31dcbe3619e0387b09621 — see commit message.

**Comment by xfgong** [comment]
@codex review — round 2 fixes pushed in a747c2da. Addressed the non-auth reason leakage into needs-action and the admin-only install endpoint mismatch in §5.1.

File: docs/superpowers/specs/2026-05-16-mcp-install-auth-handoff-spec.md [hunk 6]
```diff
@@ -0,0 +1,532 @@
...
+ ┌──────────────────────────────────────────────────────────────────────┐
+ │  ⏳  Awaiting <who>                                                   │
+ │      <who> needs to authorize this connector before it can run.      │
+ │                                                            [ Notify ] │
+ └──────────────────────────────────────────────────────────────────────┘
+```
+
+- Shown when `connector.credential_availability === 'missing'` AND the caller
+  does **not** have authority over `required_grant_scope`. The reason
+  token drives the copy:
```
[the remark this answers is not in the document]
Comment (xfgong) [reply docs/superpowers/specs/2026-05-16-mcp-install-auth-handoff-spec.md]: Fixed in 9c36ca462fa87a9594a348ebfa496074550a23a1 — narrowed §3.3 to missing_org_grant/missing_workspace_grant only.

**Comment by xfgong** [comment]
@codex review — round 3 fix pushed in 9c36ca46. Same predicate-narrowing applied to §3.3 awaiting-others band as round 2 did to §3.2 needs-action.

File: docs/superpowers/specs/2026-05-16-mcp-install-auth-handoff-spec.md [hunk 7]
```diff
@@ -0,0 +1,542 @@
...
+ ┌──────────────────────────────────────────────────────────────────────┐
+ │  ⏳  Awaiting <who>                                                   │
+ │      <who> needs to authorize this connector before it can run.      │
+ │                                                            [ Notify ] │
+ └──────────────────────────────────────────────────────────────────────┘
+```
+
+- Shown when `connector.reason` is one of `missing_org_grant |
+  missing_workspace_grant` AND the caller does **not** have authority
+  over `required_grant_scope`. As with §3.2, do NOT key on
```
[the remark this answers is not in the document]
Comment (xfgong) [reply docs/superpowers/specs/2026-05-16-mcp-install-auth-handoff-spec.md]: Fixed in 957932bd — predicate widened to include pending_oauth/grant_expired, gated on required_grant_scope∈{org,workspace}.

**Comment by xfgong** [comment]
@codex review — round 4 fix pushed in 957932bd. The §3.3 predicate now catches pending_oauth and grant_expired for org/workspace-policy installs, with per-reason copy variants.

**Comment by xfgong** [comment]
@codex review — round 4 fix pushed in https://github.com/xfgong/cubebox/commit/957932bd467fbccc3fdf637bad93e356c7e5f8b0. The §3.3 predicate now catches pending_oauth and grant_expired for org/workspace-policy installs, with per-reason copy variants.

File: docs/superpowers/specs/2026-05-16-mcp-install-auth-handoff-spec.md [hunk 8]
```diff
@@ -0,0 +1,558 @@
...
+`ApiClient`. Outputs: a Promise that resolves to `'ok' | 'cancelled' | 'error'`
+plus optional `reason`.
+
+```
+ runOAuthFlow:
+ ─────────────────────────────────────────────────────────────────────
+   1. POST <oauth/start URL>; receive { authorize_url, state }.
+   2. Open a BroadcastChannel named `cubebox-mcp-oauth`.
+   3. window.open(authorize_url, 'mcp-oauth',
+                  'width=620,height=760,noopener=no');
```
[the remark this answers is not in the document]
Comment (xfgong) [reply docs/superpowers/specs/2026-05-16-mcp-install-auth-handoff-spec.md]: Fixed in 638bec4b — pseudo-code now opens the popup synchronously on click before the await.

File: docs/superpowers/specs/2026-05-16-mcp-install-auth-handoff-spec.md [hunk 9]
```diff
@@ -0,0 +1,558 @@
...
+Already referenced by the back-end stub (`mcp_oauth.py:30-32`). This spec
+fixes its contract:
+
+- Query params accepted (all optional except `status`):
+  - `status` — `ok | error | cancelled`
+  - `install_id`
+  - `reason` — short machine-friendly token (e.g. `state_mismatch`,
+    `callback_not_wired`, `grant_write_failed`, `user_denied`)
+  - `state` — the OAuth state token, so the parent can match the message
+    to the in-flight request
```
[the remark this answers is not in the document]
Comment (xfgong) [reply docs/superpowers/specs/2026-05-16-mcp-install-auth-handoff-spec.md]: Fixed in 638bec4b — state is now required; added a sanctioned __missing__ sentinel for the genuinely-unrecoverable case.

**Comment by xfgong** [comment]
@codex review — round 5 fix pushed in 638bec4b. Synchronous popup open before the start POST, and state is now required on the return URL with an explicit unrecoverable-state sentinel.

File: docs/superpowers/specs/2026-05-16-mcp-install-auth-handoff-spec.md [hunk 10]
```diff
@@ -0,0 +1,587 @@
...
+    `state_mismatch`, `callback_not_wired`, `grant_write_failed`,
+    `user_denied`).
+
+- If the callback genuinely cannot recover `state` (corrupt URL, the
+  AS never redirected back to us), the return page MUST navigate to
+  `/oauth/mcp/return?status=error&state=__missing__&reason=state_unrecoverable`.
+  The parent matches `state === <our state>` strictly; the sentinel
+  `__missing__` is treated as "no opinion" and the parent falls
+  through to its 90 s timeout (which then closes the popup and
+  surfaces `error/timeout`). This is the only sanctioned omission of
```
[the remark this answers is not in the document]
Comment (xfgong) [reply docs/superpowers/specs/2026-05-16-mcp-install-auth-handoff-spec.md]: Fixed in 418603a5 — dropped the __missing__ sentinel entirely. Strict state match in §5.5; §5.6 renders a static fallback (no broadcast, no auto-close) for the genuinely-unrecoverable case.

**Comment by xfgong** [comment]
@codex review — round 6 fix pushed in 418603a5. Removed the sentinel that broke parallel flows; controller filter is strict, callback always carries real state.

File: docs/superpowers/specs/2026-05-16-mcp-install-auth-handoff-spec.md [hunk 11]
```diff
@@ -0,0 +1,594 @@
...
+`runOAuthFlow`. Inputs: the `oauth/start` URL (preformed per scope) and an
+`ApiClient`. Outputs: a Promise that resolves to `'ok' | 'cancelled' | 'error'`
+plus optional `reason`.
+
+```
+ runOAuthFlow (called synchronously from the user-activation click):
+ ─────────────────────────────────────────────────────────────────────
+   1. Open a blank popup synchronously, BEFORE any await:
+        const child = window.open('about:blank', 'mcp-oauth',
+                                  'width=620,height=760');
```
[the remark this answers is not in the document]
Comment (xfgong) [reply docs/superpowers/specs/2026-05-16-mcp-install-auth-handoff-spec.md]: Fixed in f942832e — target name now uses a per-flow UUID suffix.

**Comment by xfgong** [comment]
@codex review — round 7 fix pushed in f942832e. Per-flow unique popup target name.

File: docs/superpowers/specs/2026-05-16-mcp-install-auth-handoff-spec.md [hunk 12]
```diff
@@ -0,0 +1,605 @@
...
+ │     → { authorize_url, expires_at }                                 │
+ │   ↓                                                                 │
+ │  window.open(authorize_url, …)                                      │
+ │  Band transitions to `oauth-in-flight`                              │
+ │                                                                     │
+ │ Step 3 — pop-up returns                                             │
+ │  AS → /api/v1/oauth/mcp/callback?state=…&code=…                     │
+ │  Backend exchanges code, upserts MCPCredentialGrant(scope='user',   │
+ │  user_id=<the authorizing member>), redirects 302 to                │
+ │  /oauth/mcp/return?install_id=…&status=ok                           │
```
[the remark this answers is not in the document]
Comment (xfgong) [reply docs/superpowers/specs/2026-05-16-mcp-install-auth-handoff-spec.md]: Fixed in aa869a06 — flow diagrams in §5.1 and §5.3 now carry state in both the oauth/start response and the /oauth/mcp/return 302.

**Comment by xfgong** [comment]
@codex review — round 8 fix in aa869a06. State now consistently spelled out across all flow diagrams.

File: docs/superpowers/specs/2026-05-16-mcp-install-auth-handoff-spec.md [hunk 13]
```diff
@@ -0,0 +1,607 @@
...
+  - `not_installed`, `install_uninstalled`, `template_deprecated`
+    cannot reach a state where the action band is shown — the panel
+    they would render in is the install panel or an "uninstalled"
+    placeholder, both outside this band's scope.
+- `<reason copy>` translates `connector.reason` (full `MCPEffectiveReason`
+  literal for reference: `usable | not_installed | not_enabled_in_workspace
+  | install_uninstalled | template_deprecated | pending_oauth |
+  missing_org_grant | missing_workspace_grant | user_needs_connection |
+  grant_expired | discovery_failed`):
+  - `pending_oauth` → "Authorization was started but not finished."
```
[the remark this answers is not in the document]
Comment (xfgong) [reply docs/superpowers/specs/2026-05-16-mcp-install-auth-handoff-spec.md]: Fixed in c0014f4c — copy now reads as 'pending' / 'hasn't connected yet' instead of falsely implying a flow was started.

**Comment by xfgong** [comment]
@codex review — round 9 fix in c0014f4c. Copy honestly reflects pending_oauth as 'authorization is pending', not 'started but not finished'.

File: docs/superpowers/specs/2026-05-16-mcp-install-auth-handoff-spec.md [hunk 14]
```diff
@@ -0,0 +1,614 @@
...
+
+- The admin page already has a **workspace lens** (`lensWsId` in
+  `app/admin/mcp/page.tsx`). For `required=workspace`, the action band acts on
+  that lens — it surfaces and creates the workspace grant for whichever
+  workspace the admin is currently viewing. The API call still goes to the
+  workspace-scoped endpoint (`/api/v1/ws/{lensWsId}/mcp/installs/{id}/grants/workspace[…]`),
+  not to an admin alias — there is no admin alias for workspace grants and
+  this spec does not add one.
+- An admin viewing an org-policy install still sees the `org` action band on
+  the org-wide install row when `required_grant_scope === 'org'`.
```
[the remark this answers is not in the document]
Comment (xfgong) [reply docs/superpowers/specs/2026-05-16-mcp-install-auth-handoff-spec.md:271]: Fixed in 7b8dc14e — admin org row explicitly bypasses the workspace-state lens; reason derivation pinned to (install + org grant).

**Comment by xfgong** [comment]
@codex review — round 10 fix in 7b8dc14e. Admin org row now bypasses workspace-state lens; spec spells out the inputs and reason-token output.

File: docs/superpowers/specs/2026-05-16-mcp-install-auth-handoff-spec.md [hunk 15]
```diff
@@ -0,0 +1,641 @@
...
+    MCPCredentialGrant exists for this install with grant_status !=
+    'expired')`
+  - `reason` for the org row is computed from
+    `(install.auth_method, install.auth_status, presence_of_org_grant,
+    grant_status)` using the same auth-only subset of
+    MCPEffectiveReason — `pending_oauth` if auth_method='oauth' and
+    auth_status='pending' and no org grant; `missing_org_grant` if
+    auth_status='authorized' but no org grant (rotation / disconnect
+    path); `grant_expired` if the org grant is expired without
+    refresh; `usable` otherwise.
```
[the remark this answers is not in the document]
Comment (xfgong) [reply docs/superpowers/specs/2026-05-16-mcp-install-auth-handoff-spec.md]: Fixed in 86519f7d — derivation rewritten as ordered decision table; rule 5 now covers fresh static org installs where auth_status stays 'pending' until first grant.

**Comment by xfgong** [comment]
@codex review — round 11 fix in 86519f7d. Org-row reason derivation is now an ordered 5-rule decision table that handles static-org-pending and OAuth-org-rotation correctly.
